### ISSUE TITLE
[Epic]: Action possibility on each event and api request

### ISSUE THREAD

**Issue by hifabienne** [zitadel/zitadel#5101]
### **We have created a new issue for this topic. Please have a look here: https://github.com/zitadel/zitadel/issues/7235**

As a developer, I want to be able to create an action with custom code on each event and api request that can happen in ZITADEL, so that I am able to implement custom workflows.

## Acceptance criteria
- [ ] I am able to add an action to an event
- [ ] I can define when it should trigger (pre, post event)
- [ ] I am able to add an action to an api request
- [ ] I can define when the action should trigger on the request (request, response, etc)
- [ ] I am able to access the affected object of the event and be able to manipulate it in the action
- [ ] I am able to create an action that can be run on multiple events or on aggregates, e.g log to stdout on each event

https://docs.google.com/document/d/1Y4p1lXE6-4NQa7rcz6bF1PQteZ1E50ZXtDEtxowcSVk/edit#heading=h.s5bnox1o985j

### Phase 1
- [ ] #5551
- [x] #5552
- [x] #5553
- [x] #5554
- [x] #5555
- [x] #5556

### Phase 2
- [ ] Enrich context with more information (caller, command, apicall, ....) (5-10d)
       - I want to have more infos on the context, so i have them available in the actions
- [ ] Validate metadata of events (e.g. ensure that instance id is never changed, ...) (5d)
       - Event fields itself
- [ ] field validation of events on push (and first define how to validate proto validations, prepare functions, ...) (20d)
       - Validate that they dont fuck up events, fields of event itself
       - Body of event
- [ ] Document events (15-20d)
       - As a dev I need to know what events are, what events zitadel has and what they do and the body, so i am able to write my actions
       - How do we want to document? Generate?
- [ ] Implement terraform provider new action api and definition (5d)
- [ ] change examples for new action api and definition, action repository (5d)
- [ ] configure exclusions, define read only events (5d)
      - Events that should not be manipulated (e.g change password)
- [ ] end2end tests (basic testing of proper working of actions, security, ) (10-15d)
- [ ] documentation of concept (5-10d)

**Comment by hifabienne** [comment]
@eliobischof @adlerhurst Can you have a look at this, and split into sub issues so we are able to make a high estimation?

**Comment by adlerhurst** [comment]
@hifabienne what do you mean with the last criteria? what do you want to access during instance creation or org creation for example?

**Comment by adlerhurst** [comment]
as we don't compute the object as a whole the last point seems tricky to me. do you mean the write model or the object like a user?

**Comment by hifabienne** [comment]
Hmm. actually I am not sure anymore what I meant.
I list some usecase I can think of:
- If the email address changes I also want to change the username (to have them in sync)
- When the email address changes I want it to be always verified

**Comment by hifabienne** [comment]
One usecase we know already that we might should consider. 
Possibility to log each event to stdout?

**Comment by adlerhurst** [comment]
Well logging is the easy part there because we don't need additional information. 

Is the "username must be domain" setting also a valid use case? For me it currently looks like if we should consider other opportunities than passing some object additional to the event into the function. E.g. call zitadel api, add validations, add events

**Comment by hifabienne** [comment]
Yes that makes sense. I think it makes sense to have a look at some usecases and plan accordingly. 
Of course one thing is the possibility to have the actions on each event, so you can trigger an external system or something like that.
But there might be some other usecases where we need some context, data, whatever.

**Comment by adlerhurst** [comment]
> But there might be some other usecases where we need some context, data, whatever.

I totally agree, but I think bloating the parameters might be the wrong way. Adding the possibility to call the zitadel apis feels more elegant to me because users gain the flexibility to use the data they need. What we could try is to provide a zitadel client to the function or the query side, both could work out

**Comment by fforootd** [comment]
I think the best way to go is to configure a service account for each action with the associated access rights in zitadel or whatever system.

During runtime we just use that account to execute the action.

A special case is IMO if we want to allow the access to the auth api as the current user.

Besides this I would simply allow actions to be triggered on events. Even multiple in parallel. Like GH actions but only containing one script each.

**Comment by adlerhurst** [comment]
Do i get you right? You would not provide the current action triggers anymore?

**Comment by fforootd** [comment]
> Do i get you right? You would not provide the current action triggers anymore?

Correct, I think we can handle more than 90% of the cases on event basis anyways.

**Comment by adlerhurst** [comment]
> > Do i get you right? You would not provide the current action triggers anymore?
> 
> Correct, I think we can handle more than 90% of the cases on event basis anyways.

Totally agree, would shift the focus on event and aggregate design which would be worth

**Comment by fforootd** [comment]
I think one thing that potentially could be useful is to separate out pre event (command) and post event (event).

E.g I want to change claims of a token could be:

```yaml
on: 
  command: 
    - token.create
```

Or if i want do something after a token was created

```yaml
on: 
  event: 
    - token.created
```

**Comment by fforootd** [comment]
Just a thought occured.

It is hard for people to understand what commands or events exist so we either need to

A) generate docs for both
B) We could allow using the API as filter (should work with gRPC methods as well)

```yaml
on: 
  rest: 
    post:
      - /users/email
```

**Comment by adlerhurst** [comment]
another use case from discord: https://discord.com/channels/927474939156643850/1078326261098741790

It basically boils down to add additional events. In that specific case to prevent email delivery.

**Comment by adlerhurst** [comment]
> B) We could allow using the API as filter

and then you would provide the request object to the action? 

I think starting with commands/events is easier because it provides better flexibility and we can add the call to the context that users still can filter in the action

**Comment by eliobischof** [comment]
> > B) We could allow using the API as filter
> 
> and then you would provide the request object to the action?
> 
> I think starting with commands/events is easier because it provides better flexibility and we can add the call to the context that users still can filter in the action

What do you mean starting with commands/events is easier? I actually think that hooking into API calls and modifying requests instead of the lower level commands would be very understandable and wouldn't require explaining so-far-internal commands in the docs. What am I missing?

As soon as we have the login API, we also handle authorization via API, so all gaps are closed by then, AFAIU?

But @fforootd, where do you see an added value by differentiating gRPC and HTTP calls?

**Comment by mffap** [comment]
> another use case from discord: https://discord.com/channels/927474939156643850/1078326261098741790
> 
> It basically boils down to add additional commands before the commands are stored in the eventstore. In that specific case to prevent email delivery.

@adlerhurst thanks for adding this here. Just was in another call together with @fforootd and this could also apply to notifications per SMS, where the customer does not want to rely on our pre-configured provider (Twillio).
In that specific use case, their users register with their phone number and they would send an SMS with the link to passkeys enrollment via SMS, using a messaging provider's api.

**Comment by mffap** [comment]
We might need to consider allowing to use secrets / environment variables in Actions to securely re-use secrets, like API keys. The variables could maybe be set on instance level.

**Comment by adlerhurst** [comment]
> What do you mean starting with commands/events is easier? I actually think that hooking into API calls and modifying requests instead of explaining an additional so-far-internal concept like commands would be very understandable. What am I missing? Probably, also the implementation would be very easy and generic by using interceptors.

I agree that it's easier to understand that the action triggers on a request, but I think many use cases need to add/remove events as well which might get difficult on a API layer and the user needs the understanding of events/commands. We can add an interceptor layer for the eventstore as well.

**Comment by adlerhurst** [comment]
1. (2 week) Define schema for new actions
2. (1 week) Use event store interface in command
3. (1 week) Implement interceptor to manipulate events before they are written (TODO: how to validate?)
4. (2-3 week) Define flow how an action can add/manipulate events
5. (2 week) Add possibility to add events (TODO: how to validate?)
6. (2 week) Add possibility to remove events (TODO: how to validate?)
7. (2-3 week) Document events

sum: 14 weeks

**Comment by eliobischof** [comment]
My opinion:
1. (2 weeks) Move all validations from API Layer to commands
2. (3 week) Provide commands in actions API parameter for auth, mgmt and admin API
3. (1 week) Povide a consistent API for triggers
4. (1 week) Asynchronously trigger actions on all events
4. (1 week) Trigger actions on all commands before events are created
5. (3 weeks) Provide a way to manipulate events resulting from commands in actions, but in a controlled way
6. (4 weeks) Run actions before all events
8. (3 weeks) Good test coverage
4. (2 weeks) Generate documentation

sum: 20 weeks

Not included:
- Using Service Accounts with different
- Handling Secrets

**Comment by adlerhurst** [comment]
final assumption:

an action should be able to push the events analog to an interceptor:

/internal/command/user.go
func createUser
    events, err := pushInterceptor(cmds)

func pushInterceptor(cmds []eventstore.Command) ([]eventstore.Event, error)
    return action.Run(cmds)

ctx
  - command
  - api call
  - service name
  - ...

some examples:

```js
function handlerUserCreated(ctx, events) {
    events.push(new EmailSentEvent())

    writtenEvents = eventstore.push(events)
    log.log(JSON.stringify(writtenEvents))

    return writtenEvents
}
```

```js
function handlerUserCreated(events) {
    events.filter(cmd -> cmd.type === 'user.send.email')
    events = eventstore.push(events)
    log.log(JSON.stringify(events))
}
```

```js
function handlerUserCreated(events) {
    events.filter(cmd -> cmd.type === 'user.send.email')
    events = eventstore.push(events)
    log.log(JSON.stringify(events))
}
```

**Comment by mffap** [comment]
> another use case from discord: https://discord.com/channels/927474939156643850/1078326261098741790
> 
> It basically boils down to add additional commands before the commands are stored in the eventstore. In that specific case to prevent email delivery.

+1 from a discovery call. User wants to use a custom mail provider and/or email design to send mails. Payload (like codes etc.) should be delivered like a webhook. No emails should be sent through ZITADEL.

**Comment by hifabienne** [comment]
> > another use case from discord: https://discord.com/channels/927474939156643850/1078326261098741790
> > It basically boils down to add additional commands before the commands are stored in the eventstore. In that specific case to prevent email delivery.
> 
> +1 from a discovery call. User wants to use a custom mail provider and/or email design to send mails. Payload (like codes etc.) should be delivered like a webhook. No emails should be sent through ZITADEL.

With the new login api you can also tell ZITADEL to not send the email but to return the code, so you can send the emails by yourself. e.g verify email. The overarching issue can be found here: https://github.com/zitadel/zitadel/issues/5015

**Comment by mffap** [comment]
> > > another use case from discord: https://discord.com/channels/927474939156643850/1078326261098741790
> > > It basically boils down to add additional commands before the commands are stored in the eventstore. In that specific case to prevent email delivery.
> > 
> > 
> > +1 from a discovery call. User wants to use a custom mail provider and/or email design to send mails. Payload (like codes etc.) should be delivered like a webhook. No emails should be sent through ZITADEL.
> 
> With the new login api you can also tell ZITADEL to not send the email but to return the code, so you can send the emails by yourself. e.g verify email. The overarching issue can be found here: #5015

How about Password reset etc? I think this should apply for all notifications, so that one can use custom mail /sms templates as well.

**Comment by hifabienne** [comment]
> > > > another use case from discord: https://discord.com/channels/927474939156643850/1078326261098741790
> > > > It basically boils down to add additional commands before the commands are stored in the eventstore. In that specific case to prevent email delivery.
> > > 
> > > 
> > > +1 from a discovery call. User wants to use a custom mail provider and/or email design to send mails. Payload (like codes etc.) should be delivered like a webhook. No emails should be sent through ZITADEL.
> > 
> > 
> > With the new login api you can also tell ZITADEL to not send the email but to return the code, so you can send the emails by yourself. e.g verify email. The overarching issue can be found here: #5015
> 
> How about Password reset etc? I think this should apply for all notifications, so that one can use custom mail /sms templates as well.

Password reset will also be added. We just started with the email verification as this is one of the first things needed for the login flow. but we will add more and more functionality.

**Comment by mffap** [comment]
https://github.com/zitadel/zitadel/discussions/5083 reacting on the acr requirement of an application would also be an important use case to consider.

**Comment by JLedel** [comment]
I'd like to propose this feature as well when extending the action functionality: https://discord.com/channels/927474939156643850/1164888336855466025

**Comment by adlerhurst** [comment]
possibility to add the used idp to access token information: https://discord.com/channels/927474939156643850/1166523752457179238

**Comment by hifabienne** [comment]
We have sorted out the whole topic about how we want to implement out future actions.
Everything is defined in the following issue: We have created a new issue for this topic. Please have a look here: https://github.com/zitadel/zitadel/issues/7235

The issue here will be closed due to this.
